### ISSUE TITLE
ng serve removes /dist directory, but I would like to have it under version control

### ISSUE THREAD

**Issue by stepanzarubin** [angular/angular-cli#4366]
`angular-cli@1.0.0-beta.28.3`

**Comment by grizzm0** [comment]
This was done on purpose in #4293. Any reason why you would want dist in VCS?

Without a build before committing the build would be out of sync with the current revision anyhow.

**Comment by stepanzarubin** [comment]
@grizzm0 because dist is synonym of an actual build, and it is, `ng serve` serves the actual code base, but not the build, build is done once per week, month, etc. and has nothing to do with a current working directory. Another synonym of a dist is `release`, a lot of packages/libraries put their releases under `dist` directory and have it under version control.

Also, having dist under VCS allows us to keep everything in one place.

**Comment by stepanzarubin** [comment]
`ng serve` should accomplish its own task => serve actual code. And suddenly it starts to remove files from the outside? this is weird,  the other developer was confused, to get rid of a confusing it is enough to understand that there is nothing under `dist` directory when `ng serve` is running, `README` can be updated to clarify this.

At least, removing dist should not be a default behavior, can be achieved by specific configuration option or a command argument. 

Thanks.

**Comment by grizzm0** [comment]
It was changed as alot of people thought ng serve would serve the dist folder if present.

@filipesilva You got any input on this?

**Comment by stepanzarubin** [comment]
@grizzm0 then, why to have a `dist` folder at all? what is a purpose of it? temporary folder for temporary content? let`s take a look at scenario:

1. I make build: `ng build --env prod` which will write it under `dist` folder
2. Then I run `ng serve` and my build is gone... wow

Why do I stop `ng serve` at all to make a build?

**Comment by grizzm0** [comment]
Build is part of my deploy flow. I guess that's how most people use it.

**Comment by stepanzarubin** [comment]
@grizzm0 what will happen if build was not successful? for any reason, build can output files successfully, but they may not work (e.g. local build is ok, but there is something wrong with prod one).

**Comment by grizzm0** [comment]
If tests fail during deploy the deploy will be cancelled. If it builds successfully but something is wrong I would either store x releases on the prod machine or rollback by checking out a previous tag and build from that.

**Comment by stepanzarubin** [comment]
The world is not perfect, tests does not cover everything perfectly, some people do not write tests.

**Comment by grizzm0** [comment]
You could still rollback by tag or keep releases on your prod server. :p

**Comment by filipesilva** [comment]
I'm sorry this change disrupted your workflow. We're not looking at changing it back though.

`dist/` had, at best your last `ng build` and was never reflective of `ng serve`.

My recommendation for your particular case is to do make a build using the `--output-path (-op)` flag and keep that one under version control.

```
ng build --output-path=last-build/
```

**Comment by hinol** [comment]
My solution for this:

After update Jan 2017 noticed new problem. Command ng serve remove /dist dir.

Solution:

File: package.json (new command prod-build)

"scripts": {
  "ng": "ng",
  "start": "ng serve",
  "build": "ng build",
  "prod-build": "ng build --output-path=last-build/ --env=prod",
  "test": "ng test",
  "lint": "ng lint",
  "e2e": "ng e2e"
},
File: angular-cli.json (change ourDir)

"outDir": "/tmp/dist",
 
To build your prod version run: npm run prod-build

To serve: ng serve

In this configuration: ng server will create tmp dir, and your project directory will be clean, even when you commit changes without stop ng serve.

**Comment by jakeNiemiec** [comment]
Thanks @filipesilva, works like a charm.

@stepanzarubin I agree that it is bad form to delete files as a (**compulsory**) side-effect of serving my app, but I concede to @grizzm0  that (in a perfect world) I should not be committing builds to git/VCS. 

But, for some of us, it is a business requirement that we do not have control over. 

That's supposed to be the power of javascript...flexibility through composability & modularity. This "we know whats best for you" flow feel like it belongs to an old world Microsoft server, [that's why Unix won](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well).

**Comment by lgellert** [comment]
Having the latest production build in /dist was part of my build process too.   By setting up a command in package.json as @hinol suggested that sends it to a custom folder (/dist-prod/), I can live with it, minimal changes elsewhere.

My command is:
`build-prod": "ng build --prod --env=prod --output-path=dist-prod/`   note the extra --prod flag, which does the uglifying and tree shaking.

**Comment by pljspahn** [comment]
Here because even when I'm building to another output path, that path gets wiped each time I build and I lose the .git folder.

Why can't it just spit the output to that folder instead of removing everything and then dumping the contents there?

**Comment by jakeNiemiec** [comment]
>Build is part of my deploy flow. I guess that's how most people use it.

@grizzm0 I need to flip on this issue. Turns out I wasn't giving this a fair chance. Having TravisCI build & publish my npm package is so much better than commiting /dist to VCS.

**Comment by JasonGoemaat** [comment]
This is SOOO wrong.  'ng serve' is deleting an entire directory's contents that might not even be your build output if you've changed its location.  There is no documentation on this, deleting an entire directory seems to be a 'hidden feature' of the 'ng serve' command.  In my experience having programs delete a directory without informing the user is a bad idea.  At the **very least** there should be an output line when running 'ng serve' that says `Deleting the contents of the dist directory...`

**Comment by filipesilva** [comment]
I should mention that on `1.1.0-rc.0` there is a new `--delete-output-path` flag, defaulting to false. We introduced it both for Docker setups that link the dist directory, and for custom setups that use another build system to produce artifacts in dist.

**Comment by dohpaz42** [comment]
Hello. @filipesilva and/or @grizzm0, I'm a bit new to the front-end development and using Angular's cli. I've never done deployments before, so I'm curious why having a `dist/` folder that's not empty is somehow a bad thing?

Also, isn't the suggestion of renaming the `dist/` folder (i.e., `last-build/`) nothing more than lipstick on the proverbial pig? In other words, whatever the problem is with having a `dist/` folder won't actually be solved, because you've simply renamed it. 

The reason why it confuses me is because I was reading [an article](https://medium.com/@cyrilletuzi/how-to-build-and-publish-an-angular-module-7ad19c0b4464#b447) about creating angular modules for NPM, and it's rule of thumb is to build into the dist directory, and push to npm from there, so that only the necessary files get published. If this is not the correct workflow, would you have any documentation about the best practice workflow developers should be using?

Thank you in advance!

**Comment by filipesilva** [comment]
@dohpaz42 while a `dist/` folder contains build artifacts for both apps and libraries, they are used differently. For apps, the dist folder is often deployed, while for libraries it can be installed via the dependency manager or linked for live development. 

There is no problem in having a non-empty `dist/` per se, but due to it being a auto-generated folder in both cases, it's content is transient during development cycles. The build system deletes and recreates it's contents as needed.

Take for instance fingerprinting - the hash that gets appended during production builds. If `dist/` wasn't deleted wholesale between builds, it would quickly accumulate a lot of past artifacts that are not used anymore. This also happens for any files in assets that you had at one point and removed since.

There are some advanced usecases where it is desirable to not delete the `dist/` folder, such as symlinks for docker development or when multiple build systems are outputting to `dist/`, but those are the exceptions rather than the norm. 

Angular CLI does not yet support libraries (https://github.com/angular/angular-cli/issues/1692), but you can find some temporary guidance in https://github.com/angular/angular-cli/issues/1692#issuecomment-285363067 and https://github.com/angular/angular-cli/issues/1692#issuecomment-301736484.

**Comment by severin2** [comment]
I realize that this is closed, but I keep ending up on this issue while searching for a while to make ng-cli clean everything inside of dist/ but not remove the folder. There's a definite use case for it.

In our case we are building the app with the angular cli and serving it with the nginx docker container. During development, we simply volume map the dist directory of our project to that container so we can see updates without starting the container.

Docker volume mapping strongly disagrees with removing the root volume mapped directory. And yeah we could map the next directory up and use docker trickery to ignore large folders like node_modules.

As another point of interest, we have a separate project where we added webpack manually to an old app, and we had to explicitly tell it to clean dist/* insead of dist/ to solve this problem.

**Comment by filipesilva** [comment]
@severin2 the `--delete-output-path=false` option was added for that specific scenario. See https://github.com/angular/angular-cli/wiki/build for details.

**Comment by jayoungers** [comment]
I believe `ng serve --delete-output-path false` is working for me.

One thing to note that while the flag is documented in the wiki for the build command, it's not documented for serve.

In my situation, once deployed the UI and API run under the same webserver.  As part of the build I have a test to verify static content is being served with a no-cache header.  That works fine on the build server, but locally I wasn't able to run the test since the dist folder was being removed on `ng serve` (npm postinstall will stub in a fake file specifically for this test; during the actual build that folder gets replaced)

**Comment by izy2nv** [comment]
@filipesilva I'm trying to integrate my angular 4 project in an already existing maven project. I understand the '/dist' folder has to be referenced in the 'target' dir of my maven project. How can this integratin be done in this case since angular-cli deletes my dist folder?

**Comment by bastienJS** [comment]
@izy2nv  Try ``` ng serve --deleteOutputPath=false ``` that worked for me prevention the deletion of my outDir value.

**Comment by felyperennan** [comment]
Its very lame that Ng Serve deletes everything on release path to this day. I got around with making a custom npm run build to use another output path, but its still lame as fuck that Angular messes up development and production output in this ridiculous manner.

**Comment by dohpaz42** [comment]
@felyperennan open a new ticket. this one is closed, and as such, it seems the devs stop caring about it.

**Comment by felyperennan** [comment]
I'm not asking this to be changed, I think everyone already adapted to this by this point, I'm just adding how a custom npm script can be even more beneficial than having this particular behavior changed.

**Comment by johnking** [comment]
Just want to add my two cents here. 

I am developing a desktop app with `electron`, so I need to keep the `dist` folder to make electron running, in the meantime, I can run `ng server` for angular component development.  

` ng serve --deleteOutputPath=false` is good enough for my case. 

thanks

**Comment by shoagraw** [comment]
it deleted some on my work in progress file and not able to find those files now, not even in recycle bin,
Any idea where these files will go after ng build deletes these.

**Comment by jakeNiemiec** [comment]
@shoagraw, they are deleted permanently from the filesystem: https://github.com/angular/angular-cli/blob/master/tools/publish/src/build.ts#L75

**Comment by hamdimefteh** [comment]
change /dist to other folder path !

**Comment by Terrafire123** [comment]
Running "ng build --output-path='../'" , instead of creating a directory in the parent named "dist", just deleted 12 hours of my work.

A little **warning**, or maybe an error message would have been nice.

Instead, AFTER my entire project was deleted, I got a cheery little message saying "Path must be a string. Received null"

**Comment by michaelb-01** [comment]
Has the delete-output-path flag been removed in Angular CLI 6.0.0?

**Comment by rsheptolut** [comment]
It also seems Angular CLI 6.0.0 stopped deleting the "dist" folder by default. Had to add `rmdir .\\dist /S /Q && ` to my npm `start` script before `ng serve`, because oddly enough I've depended on the old behavior 🙂

**Comment by jakeNiemiec** [comment]
Looks like they learned that side-effects are bad (however well-intentioned):

Before: https://github.com/angular/angular-cli/pull/4293
Code: https://github.com/filipesilva/angular-cli/blob/4a653e2820539cf71faf51ac2e0e195c23a234bd/packages/angular-cli/tasks/serve.ts

After: https://github.com/angular/angular-cli/pull/9883
Code: https://github.com/angular/angular-cli/blob/bc56efa0535fe068b976fffaf3c2e570c967a333/packages/%40angular/cli/commands/serve.ts
